x-pack/plugins/security_solution/scripts/run_cypress/parallel_serverless.ts

```diff
@@ -0,0 +1,531 @@
+import globby from 'globby';
+import pMap from 'p-map';
+import { ToolingLog } from '@kbn/tooling-log';
+import { withProcRunner } from '@kbn/dev-proc-runner';
+import cypress from 'cypress';
+import { findChangedFiles } from 'find-cypress-specs';
+import grep from '@cypress/grep/src/plugin';
+import crypto from 'crypto';
+import fs from 'fs';
+import { createFailError } from '@kbn/dev-cli-errors';
+import axios from 'axios';
+import path from 'path';
+import os from 'os';
+import { renderSummaryTable } from './print_run';
+import type { SecuritySolutionDescribeBlockFtrConfig } from './utils';
+import { isSkipped, parseTestFileConfig } from './utils';
+
+interface ProductType {
+  product_line: string;
+  product_tier: string;
+}
+
+interface CreateEnvironmentRequestBody {
+  name: string;
+  region_id: string;
+  product_types: ProductType[];
+}
+
+interface Environment {
+  name: string;
+  id: string;
+  region: string;
+  es_url: string;
+  kb_url: string;
+  product: string;
+}
+
+interface Credentials {
+  username: string;
+  password: string;
+}
+
+const DEFAULT_REGION = 'aws-eu-west-1';
+let log: ToolingLog;
+
+/**
+ * Retrieve test files using a glob pattern.
+ * If process.env.RUN_ALL_TESTS is true, returns all matching files, otherwise, return files that should be run by this job based on process.env.BUILDKITE_PARALLEL_JOB_COUNT and process.env.BUILDKITE_PARALLEL_JOB
+ */
+const retrieveIntegrations = (integrationsPaths: string[]) => {
+  const nonSkippedSpecs = integrationsPaths.filter((filePath) => !isSkipped(filePath));
+
+  if (process.env.RUN_ALL_TESTS === 'true') {
+    return nonSkippedSpecs;
+  } else {
+    // The number of instances of this job were created
+    const chunksTotal: number = process.env.BUILDKITE_PARALLEL_JOB_COUNT
+      ? parseInt(process.env.BUILDKITE_PARALLEL_JOB_COUNT, 10)
+      : 1;
+    // An index which uniquely identifies this instance of the job
+    const chunkIndex: number = process.env.BUILDKITE_PARALLEL_JOB
+      ? parseInt(process.env.BUILDKITE_PARALLEL_JOB, 10)
+      : 0;
+
+    const nonSkippedSpecsForChunk: string[] = [];
+
+    for (let i = chunkIndex; i < nonSkippedSpecs.length; i += chunksTotal) {
+      nonSkippedSpecsForChunk.push(nonSkippedSpecs[i]);
+    }
+
+    return nonSkippedSpecsForChunk;
+  }
+};
+
+const encode = (str: string): string => Buffer.from(str, 'binary').toString('base64');
+
+const getApiKeyFromElasticCloudJsonFile = () => {
+  const userHomeDir = os.homedir();
+  try {
+    const jsonString = fs.readFileSync(path.join(userHomeDir, '.elastic/cloud.json'), 'utf-8');
+    const jsonData = JSON.parse(jsonString);
+    return jsonData.api_key.qa;
+  } catch (e) {
+    log.info('API KEY could not be found in ');
+    return null;
+  }
+};
+
+// Poller function that is polling every 20s, forever until function is resolved.
+async function poll<T>(
+  fn: () => Promise<T>,
+  retries: number = 200,
+  interval: number = 20000
+): Promise<T> {
+  return Promise.resolve()
```

Comment:
@dkirchan this is nice you implemented poll function here. I'm afraid it's additional maintenance burden on us. If we could reuse something existing it would make our life easier 😄  Have you considered using `p-retry` instead? It does exponential backoff which may work even better.